### PR TITLE
feat: add support for markdown code block

### PR DESCRIPTION
**Pull request by wa0x6e**
### Summary

Closes: #249 

Add support for markdown code block

Screenshot 2024-04-07 at 20 45 47

### How to test

1. Go to http://localhost:8080/#/s:test.wa0x6e.eth/proposal/0xf56083e1fcbdc5c9ad15d9e91f67a72b817bdf83f1589af8db20e27385db95e1
2. It should parse and render the code block properly

### Todo

- [x] Add copy code button for code block
- [x] Add syntax highlighting for code block

### CODE REVIEW

**Comment by bonustrack** [comment]
Is there an easy way to support color in syntax based on lang, similar than GitHub:

ts:
```ts
const a = 'b';
```

json:
```json
{
  "wan": "hello world"
}
```

without lang:
```
{
  "wan": "hello world"
}
```

**Comment by bonustrack** [comment]
Copy button is great too

**Comment by wa0x6e** [comment]
Copy button + syntax highlighting added.

File: apps/ui/src/components/Ui/Markdown.vue [hunk 1]
```diff
@@ -1,21 +1,39 @@
...
   html: false,
   breaks: true,
   typographer: false,
-  linkTarget: '_blank'
+  linkTarget: '_blank',
+  highlight: function (str, lang) {
+    if (lang && hljs.getLanguage(lang)) {
+      try {
+        return hljs.highlight(str, { language: lang }).value;
+      } catch (err) {}
```
Comment (Sekhmet) [inline apps/ui/src/components/Ui/Markdown.vue:25]: nit. we migrated to using `e` as a name for error in catch block (we have one left it looks like though :D)

File: apps/ui/src/components/Ui/Markdown.vue [hunk 2]
```diff
@@ -36,13 +54,42 @@ const parsed = computed(() => {
 
   return remarkable.render(formattedBody);
 });
+
+onMounted(() => {
+  const body = document.querySelector('.markdown-body');
+  if (body !== null) {
```
Comment (Sekhmet) [inline apps/ui/src/components/Ui/Markdown.vue:60]: If we just flip it then we can do `if (!body) return` and code below is cleaner.

File: apps/ui/src/components/Ui/Markdown.vue [hunk 3]
```diff
@@ -36,13 +54,42 @@ const parsed = computed(() => {
...
+    body.querySelectorAll('pre>code').forEach(function (code) {
```
Comment (Sekhmet) [inline apps/ui/src/components/Ui/Markdown.vue:61]:
```suggestion
    body.querySelectorAll('pre>code').forEach((code) => {
```

File: apps/ui/src/components/Ui/Markdown.vue [hunk 4]
```diff
@@ -36,13 +54,42 @@ const parsed = computed(() => {
...
+      const parent = code.parentElement;
+      const copyButton = document.createElement('button');
+
+      copyButton.classList.add('copy');
+      copyButton.setAttribute('type', 'button');
+      copyButton.innerHTML = `<svg viewBox="0 0 24 24" width="20px" height="20px">${icons.icons.duplicate.body}</svg>`;
+
+      copyButton.addEventListener('click', function () {
```
Comment (Sekhmet) [inline apps/ui/src/components/Ui/Markdown.vue:69]:
```suggestion
      copyButton.addEventListener('click', () => {
```

**Comment by bonustrack** [comment]
Works well, there is just this highlight.js that is quite heavy, is there any way to make it lighter?

Before:
image
After:
image

**Comment by Sekhmet** [comment]
What if we limit languages we load?
```js
import hljs from 'highlight.js/lib/core';
import javascript from 'highlight.js/lib/languages/javascript';
hljs.registerLanguage('javascript', javascript);
```

We should come up with list of languages we want to support so we don't bring extras.

**Comment by bonustrack** [comment]
JS, TS, JSON, solidity? Python and Rust maybe, I think thats it

**Comment by wa0x6e** [comment]
Done, can you recheck size with you tool ? I skipped Solidity, as it's requiring a 3rd party plugin

**Comment by bonustrack** [comment]
You can check by yourself after build its here: https://snapshotx.xyz/stats.html

**Comment by wa0x6e** [comment]
Getting this

Screenshot 2024-04-10 at 13 04 00

**Comment by bonustrack** [comment]
Looks good, would that work if I use "ts" or "js" instead of "javascript"?

File: apps/ui/src/components/Ui/Markdown.vue [hunk 5]
```diff
@@ -1,21 +1,48 @@
...
 import { computed } from 'vue';
 import { Remarkable } from 'remarkable';
+import { icons } from '@iconify-json/heroicons-outline';
 import { linkify } from 'remarkable/linkify';
 import { getUrl } from '@/helpers/utils';
 
+hljs.registerLanguage('javascript', javascript);
+hljs.registerLanguage('json', json);
+hljs.registerLanguage('rust', rust);
+hljs.registerLanguage('python', python);
```
Comment (Sekhmet) [inline apps/ui/src/components/Ui/Markdown.vue:16]: Does highlight.js support Solidity? Solidity was suggested by less and it's missing there.

Comment (wa0x6e) [reply apps/ui/src/components/Ui/Markdown.vue:16]: It does not, it requires this plugin https://github.com/highlightjs/highlightjs-solidity

Should we proceed with this additional dependency ?

Comment (Sekhmet) [reply apps/ui/src/components/Ui/Markdown.vue:16]: Looks like it works with older version of highlight (v9).

Comment (wa0x6e) [reply apps/ui/src/components/Ui/Markdown.vue:16]: v9 is pre-covid era

**Comment by wa0x6e** [comment]
> Looks good, would that work if I use "ts" or "js" instead of "javascript"?

Seems to work, even with just ` ``` ` , highlight.js have a language detection

File: apps/ui/src/components/Ui/Markdown.vue (hunk 5 above)
Comment (bonustrack) [reply apps/ui/src/components/Ui/Markdown.vue:16]: Let's skip Solidity for now, not a big deal

Comment (wa0x6e) [reply apps/ui/src/components/Ui/Markdown.vue:16]: It will fallback and be parsed as javascript I think

**Review by Sekhmet — APPROVED** [review]
tACK

**Pull request merged**